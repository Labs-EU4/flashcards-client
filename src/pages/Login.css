.login-form {
    max-width: 400px;
    margin: 0 auto;
  }
  .login-form-forgot {
    float: right;
  }
  .login-form-button {
    width: 100%;
  }
  .ant-form-explain {
      color: red;
  }
  .forgot-password h1{
      margin: 0 auto;
      text-align: center;
      margin-bottom: 30px;
      font-size: 2rem;
  }