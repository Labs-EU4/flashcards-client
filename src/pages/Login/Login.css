.login-form {
  max-width: 400px;
  margin: 0 auto;
}
.login-form-forgot {
  float: right;
}
.login-form-button {
  width: 100%;
}
.ant-form-explain {
  color: red;
}
.forgot-password h1 {
  margin: 0 auto;
  text-align: center;
  margin-bottom: 30px;
  font-size: 2rem;
}
.login-container {
  position: fixed;
  display: flex;
  flex-direction: column;
  width: 100%;
  align-items: center;
  margin: 30vh auto;
}
h1 {
  display: flex;
  justify-content: center;
}
.google-auth {
  width: 100%;
  display: flex;
  border: 1px solid #d9d9d9;
  border-radius: 4px;
  align-items: center;
  justify-content: center;
}

.google-icon {
  width: 1.8rem;
  height: 1.8rem;
  margin: 0 10px;
}
