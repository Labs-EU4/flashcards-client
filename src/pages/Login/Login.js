import React, {useState, useEffect} from "react";
import {connect} from "react-redux";
import {Link} from "react-router-dom";
import {Form, Icon, Input, Button, Spin, Alert} from "antd";
import GoogleButton from "../../components/GoogleLogin/GoogleButton";
import {login} from "../../state/actions/auth";
import * as styles from "./Login.module.css";
import backgroundStyles from "../../components/formStyleComponent/FormStyleComponent.module.css";
import FormHeader from "../../components/formStyleComponent/FormHeader";
import {axiosWithAuth} from "../../utils/axios";

export const Login = props => {
  const {
    getFieldDecorator,
    getFieldsError,
    validateFields,
    isFieldTouched,
    getFieldError,
    setFieldsValue,
  } = props.form;
  const [formValues, setFormValues] = useState({
    email: "",
    password: "",
  });
  const [error, setError] = useState("");
  const [loading, setLoading] = useState(false);
  const hasErrors = fieldsError => {
    return Object.keys(fieldsError).some(field => fieldsError[field]);
  };
  const handleChange = e => {
    setFormValues({
      ...formValues,
      [e.target.name]: e.target.value,
    });
    setFieldsValue({[e.target.name]: e.target.value});
  };

  const handleSubmit = async e => {
    e.preventDefault();
    setLoading(true);

    try {
      await props.login(formValues);
      props.history.push("/");
    } catch (error) {
      setError(error.response.data.message || "Something went wrong");
    } finally {
      setLoading(false);
      props.form.resetFields();
    }
  };

  useEffect(() => {
    async function checkIfTokenValid() {
      if (localStorage.getItem("token")) {
        try {
          const response = await axiosWithAuth().post("/auth/validate_token");
          if (response.status === 201) {
            props.history.push("/");
          }
        } catch (error) {
          console.log("Token invalid or expired, please login again!");
        }
      }
    }
    checkIfTokenValid();
    validateFields({suppressWarning: true});
  }, [props.history, validateFields]);
  const emailError = isFieldTouched("email") && getFieldError("email");
  const passwordError = isFieldTouched("password") && getFieldError("password");
  return (
    <div className={backgroundStyles.formStyle}>
      <FormHeader />
      <div>
        <h1>Login</h1>
        <Spin spinning={loading}>
          <Form onSubmit={handleSubmit} className="login-form">
            <Form.Item
              validateStatus={emailError ? "error" : ""}
              hasFeedback
              help={emailError || ""}
            >
              {getFieldDecorator("email", {
                //rules are for the form validation
                rules: [
                  {required: true, message: "Please input a valid email!"},
                  {
                    type: "email",
                    message: "Invalid email",
                  },
                ],
              })(
                <Input
                  data-testid="email-input"
                  name="email"
                  onInput={handleChange}
                  //form icon in the email field, change type for different icons, see antdesign docs
                  prefix={<Icon type="mail" style={{color: "rgba(0,0,0,.25)"}} />}
                  placeholder="Email"
                />
              )}
            </Form.Item>
            <Form.Item
              validateStatus={passwordError ? "error" : ""}
              help={passwordError || ""}
              hasFeedback
            >
              {getFieldDecorator("password", {
                //rules are for the form validation
                rules: [{required: true, message: "Please input a password!"}],
              })(
                <Input
                  name="password"
                  type="password"
                  data-testid="password-input"
                  onInput={handleChange}
                  //form icon in the email field, change type for different icons, see antdesign docs
                  prefix={<Icon type="lock" style={{color: "rgba(0,0,0,.25)"}} />}
                  placeholder="Password"
                />
              )}
            </Form.Item>
            <Form.Item>
              <Button
                type="primary"
                data-testid="login-button"
                htmlType="submit"
                className={styles.login_button}
                disabled={hasErrors(getFieldsError())}
                loading={loading}
              >
                Login
              </Button>
              Forgot password? <Link to="/reset-password">click here to reset! </Link>
              Or <Link to="/register">register here!</Link>
              <GoogleButton>SIGN IN WITH GOOGLE</GoogleButton>
            </Form.Item>
            {error ? (
              <Alert
                message={error}
                type="error"
                data-testid="server-alert"
                closable
                afterClose={() => setError(null)}
              />
            ) : null}
          </Form>
        </Spin>
      </div>
    </div>
  );
};

export const WrappedNormalLoginForm = Form.create({name: "normal_login"})(Login);

export default connect(null, {login})(WrappedNormalLoginForm);
