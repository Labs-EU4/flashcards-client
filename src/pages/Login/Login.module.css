.login_button {
  width: 100%;
}
