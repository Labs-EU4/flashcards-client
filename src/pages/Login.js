import React, {useState} from "react";
import {Form, Icon, Input, Button, Spin, Alert} from "antd";
import "../pages/Login.css";
import {connect} from "react-redux";
import {login} from "../state/IsLoggedIn/IsLoggedInActionCreators";

export const Login = props => {
  const [formValues, setFormValues] = useState({
    email: "",
    password: "",
  });
  const [error, setError] = useState("");
  const [loading, setLoading] = useState(false);
  const handleChange = e => {
    setFormValues({
      ...formValues,
      [e.target.name]: e.target.value,
    });
  };
  const handleSubmit = async e => {
    e.preventDefault();
    setLoading(true);
    try {
      await props.login(formValues);
    } catch (error) {
      setError(error);
      console.log(error);
    } finally {
      setLoading(false);
    }
  };
  const {getFieldDecorator} = props.form;
  return (
    <div className="login-container">
      <Spin spinning={loading} delay={300}>
        <h1>Login</h1>
        <Form onSubmit={handleSubmit} className="login-form">
          <Form.Item>
            {getFieldDecorator("email", {
              //rules are for the form validation
              rules: [
                {required: true, message: "Please input a valid email!"},
                {
                  type: "email",
                  message: "Invalid email",
                },
              ],
            })(
              <Input
                name="email"
                setFieldsValue={formValues.email}
                onChange={handleChange}
                //form icon in the email field, change type for different icons, see antdesign docs
                prefix={<Icon type="mail" style={{color: "rgba(0,0,0,.25)"}} />}
                placeholder="Email"
              />
            )}
          </Form.Item>
          <Form.Item>
            {getFieldDecorator("password", {
              //rules are for the form validation
              rules: [
                {required: true, message: "Please input a password!"},
                {
                  type: "string",
                  message: "Invalid password",
                },
              ],
            })(
              <Input
                name="password"
                type="password"
                setFieldsValue={formValues.password}
                onChange={handleChange}
                //form icon in the email field, change type for different icons, see antdesign docs
                prefix={<Icon type="lock" style={{color: "rgba(0,0,0,.25)"}} />}
                placeholder="Password"
              />
            )}
          </Form.Item>
          <Form.Item>
            <Button type="primary" htmlType="submit" className="login-form-button">
              Login
            </Button>
            Or <a href="/register">Register Here!</a>
          </Form.Item>
          {error && (
            <Alert
              message={error}
              type="error"
              closable
              afterClose={() => setError(null)}
            />
          )}
        </Form>
      </Spin>
    </div>
  );
};

export const WrappedNormalLoginForm = Form.create({name: "normal_login"})(Login);

export default connect(() => {}, {login})(WrappedNormalLoginForm);
