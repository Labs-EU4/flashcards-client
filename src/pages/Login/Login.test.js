import React from "react";
import * as rtl from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import {BrowserRouter} from "react-router-dom";
import {WrappedNormalLoginForm as Login} from "./Login";

let promise;
const mockLogin = jest.fn();
const mockHistory = {push: jest.fn()};
let wrapper;
beforeEach(() => {
  jest.resetAllMocks();
  promise = Promise.resolve();
  wrapper = rtl.render(
    <BrowserRouter>
      <Login
        login={mockLogin.mockImplementationOnce(() => promise)}
        history={mockHistory}
      />
    </BrowserRouter>
  );
});

const EmailInput = () => {
  return wrapper.getByTestId("email-input");
};

const PasswordInput = () => {
  return wrapper.getByTestId("password-input");
};

const Button = () => {
  return wrapper.getByTestId("login-button");
};

it("renders without crashing", () => {
  expect(wrapper.container).toBeInTheDocument();
});

it("renders all expected elements", () => {
  expect(EmailInput()).toBeInTheDocument();
  expect(PasswordInput()).toBeInTheDocument();
  expect(Button()).toBeInTheDocument();
});

it("does not submit due to submit button being disabled", () => {
  expect(Button()).toBeDisabled();
});

it("does not make an axios call if some input fields are left blank", () => {
  rtl.fireEvent.click(Button());
  expect(mockLogin).toHaveBeenCalledTimes(0);
});

it("Calls correct action on submit", async () => {
  rtl.fireEvent.change(EmailInput(), {
    target: {value: "darragh@test.com"},
  });
  rtl.fireEvent.change(PasswordInput(), {
    target: {value: "123456789"},
  });
  rtl.fireEvent.click(Button());
  await rtl.act(() => promise);
  expect(mockLogin).toHaveBeenCalledTimes(1);
  expect(mockLogin.mock.calls[0]).toMatchInlineSnapshot(`
    Array [
      Object {
        "email": "",
        "password": "",
      },
    ]
  `);
});
