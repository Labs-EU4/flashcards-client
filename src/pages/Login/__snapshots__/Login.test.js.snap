// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div>
  <div
    class="formStyle"
  >
    <div
      class="header"
    >
      <a
        href="/"
      >
        <img
          alt="logo"
          src="https://i.imgur.com/tuS7kwh.png"
          width="60px"
        />
      </a>
    </div>
    <div>
      <h1>
        Login
      </h1>
      <form
        class="ant-form ant-form-horizontal login-form"
      >
        <div
          class="ant-row ant-form-item"
        >
          <div
            class="ant-col ant-form-item-control-wrapper"
          >
            <div
              class="ant-form-item-control"
            >
              <span
                class="ant-form-item-children"
              >
                <span
                  class="ant-input-affix-wrapper"
                >
                  <span
                    class="ant-input-prefix"
                  >
                    <i
                      aria-label="icon: mail"
                      class="anticon anticon-mail"
                      style="color: rgba(0, 0, 0, 0.25);"
                    >
                      <svg
                        aria-hidden="true"
                        class=""
                        data-icon="mail"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M928 160H96c-17.7 0-32 14.3-32 32v640c0 17.7 14.3 32 32 32h832c17.7 0 32-14.3 32-32V192c0-17.7-14.3-32-32-32zm-40 110.8V792H136V270.8l-27.6-21.5 39.3-50.5 42.8 33.3h643.1l42.8-33.3 39.3 50.5-27.7 21.5zM833.6 232L512 482 190.4 232l-42.8-33.3-39.3 50.5 27.6 21.5 341.6 265.6a55.99 55.99 0 0 0 68.7 0L888 270.8l27.6-21.5-39.3-50.5-42.7 33.2z"
                        />
                      </svg>
                    </i>
                  </span>
                  <input
                    class="ant-input"
                    data-__field="[object Object]"
                    data-__meta="[object Object]"
                    data-testid="email-input"
                    id="normal_login_email"
                    name="email"
                    placeholder="Email"
                    setfieldsvalue=""
                    type="text"
                    value=""
                  />
                </span>
              </span>
            </div>
          </div>
        </div>
        <div
          class="ant-row ant-form-item"
        >
          <div
            class="ant-col ant-form-item-control-wrapper"
          >
            <div
              class="ant-form-item-control"
            >
              <span
                class="ant-form-item-children"
              >
                <span
                  class="ant-input-affix-wrapper"
                >
                  <span
                    class="ant-input-prefix"
                  >
                    <i
                      aria-label="icon: lock"
                      class="anticon anticon-lock"
                      style="color: rgba(0, 0, 0, 0.25);"
                    >
                      <svg
                        aria-hidden="true"
                        class=""
                        data-icon="lock"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M832 464h-68V240c0-70.7-57.3-128-128-128H388c-70.7 0-128 57.3-128 128v224h-68c-17.7 0-32 14.3-32 32v384c0 17.7 14.3 32 32 32h640c17.7 0 32-14.3 32-32V496c0-17.7-14.3-32-32-32zM332 240c0-30.9 25.1-56 56-56h248c30.9 0 56 25.1 56 56v224H332V240zm460 600H232V536h560v304zM484 701v53c0 4.4 3.6 8 8 8h40c4.4 0 8-3.6 8-8v-53a48.01 48.01 0 1 0-56 0z"
                        />
                      </svg>
                    </i>
                  </span>
                  <input
                    class="ant-input"
                    data-__field="[object Object]"
                    data-__meta="[object Object]"
                    data-testid="password-input"
                    id="normal_login_password"
                    name="password"
                    placeholder="Password"
                    setfieldsvalue=""
                    type="password"
                    value=""
                  />
                </span>
              </span>
            </div>
          </div>
        </div>
        <div
          class="ant-row ant-form-item"
        >
          <div
            class="ant-col ant-form-item-control-wrapper"
          >
            <div
              class="ant-form-item-control"
            >
              <span
                class="ant-form-item-children"
              >
                <button
                  class="ant-btn login-form-button ant-btn-primary"
                  data-testid="login-button"
                  disabled=""
                  type="submit"
                >
                  <span>
                    Login
                  </span>
                </button>
                Forgot password? 
                <a
                  href="/reset-password"
                >
                  click here to reset! 
                </a>
                Or 
                <a
                  href="/register"
                >
                  register here!
                </a>
                <a
                  class="google-auth"
                  href="https://flashdecks.herokuapp.com/api/auth/google"
                >
                  <img
                    alt="google-icon"
                    class="google-icon"
                    src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
                  />
                  SIGN IN WITH GOOGLE
                </a>
              </span>
            </div>
          </div>
        </div>
      </form>
    </div>
  </div>
</div>
`;
